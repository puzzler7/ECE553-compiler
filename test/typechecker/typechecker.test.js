const fs = require('fs');
const { spawnSync, spawn } = require('child_process');
const { test, it } = require('@jest/globals');

const cmMakeMsg = '[New bindings added.]';
const errorMessagePattern = /tig:[0-9]+.[0-9]+:/;
const positiveTests = [
    'merge.tig',
    'queens.tig',
    'test1.tig',
    'test2.tig',
    'test3.tig',
    'test4.tig',
    'test5.tig',
    'test6.tig',
    'test7.tig',
    'test8.tig',
    'test12.tig',
    'test27.tig',
    'test30.tig',
    'test37.tig',
    'test41.tig',
    'test42.tig',
    'test44.tig',
    'test46.tig',
    'test47.tig',
    'test48.tig'
]

const getRelativePathOfTests = (dir) => {
    return fs.readdirSync(dir)
             .filter(fileName => /^.*\.tig$/.exec(fileName))
             .map(fileName => dir + fileName);

}

const getScript = (filePath) => {
    return `CM.make "src/typechecker/sources.cm"; Main.typecheckOnly("${filePath}");`
}

describe('The textbook tests', () => {
    const cases = getRelativePathOfTests('test/testcases/');
    const positiveCases = cases.filter(path => 
        positiveTests.find(test => path.includes(test))
    )
    const negativeCases = cases.filter(path => !positiveCases.includes(path))

    test.each(positiveCases)(
        'finds no type errors in %s',
        (testFile) => {
            const parseResults = spawnSync(`sml`, { input: getScript(testFile) })
            expect(parseResults.error).toBeFalsy();
            expect(parseResults.stderr.toString()).toBe('');

            const stdout = parseResults.stdout.toString()
            expect(stdout.includes(cmMakeMsg)).toBeTruthy();
            expect(stdout.slice(stdout.indexOf(cmMakeMsg) + cmMakeMsg.length)).toMatchSnapshot();
            expect(errorMessagePattern.test(parseResults.stdout.toString())).toBeFalsy();
        }
    );

    test.each(negativeCases)(
        'correctly finds type errors in %s',
        (testFile) => {
            const parseResults = spawnSync(`sml`, { input: getScript(testFile) })
            expect(parseResults.error).toBeFalsy();
            expect(parseResults.stderr.toString()).toBe('');

            const stdout = parseResults.stdout.toString()
            expect(stdout.includes(cmMakeMsg)).toBeTruthy();
            expect(stdout.slice(stdout.indexOf(cmMakeMsg) + cmMakeMsg.length)).toMatchSnapshot();
            expect(errorMessagePattern.test(parseResults.stdout.toString())).toBeTruthy();
        }
    );
})