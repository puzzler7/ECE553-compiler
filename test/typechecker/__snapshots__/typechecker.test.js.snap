// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The textbook tests parses test/testcases/merge.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/queens.tig correctly 1`] = `
"
val it = true : bool
intintArrayintArrayintArrayintArray
uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test1.tig correctly 1`] = `
"
val it = true : bool
intarrtypeval it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test2.tig correctly 1`] = `
"
val it = true : bool
 namesymbol:intmyintarrtypeval it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test3.tig correctly 1`] = `
"
val it = true : bool
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test4.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test5.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test5.tig:4.26:Type undefined
test/testcases/test5.tig:7.23:Type undefined
test/testcases/test5.tig:8.28:Type undefined
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test6.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test7.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test8.tig correctly 1`] = `
"
val it = true : bool
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test9.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test9.tig:3.1:Then Else disagree
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test10.tig correctly 1`] = `
"
val it = true : bool
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test11.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test11.tig:2.1:Not int type

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:58.71
- 
"
`;

exports[`The textbook tests parses test/testcases/test12.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:58.71
- 
"
`;

exports[`The textbook tests parses test/testcases/test13.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test13.tig:3.3:Not int type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test14.tig correctly 1`] = `
"
val it = true : bool
intarrtypetest/testcases/test14.tig:12.9:Not int type
test/testcases/test14.tig:12.9:Not int type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test15.tig correctly 1`] = `
"
val it = true : bool
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test16.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test16.tig:4.8:type does not exist
 namesymbol:atest/testcases/test16.tig:6.8:type does not exist
 namesymbol:aval it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test17.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test17.tig:4.23:Type undefined
test/testcases/test17.tig:6.28:Type undefined
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test18.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test19.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test20.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test20.tig:3.18:Variable does not exist!
test/testcases/test20.tig:3.19:Not int type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test21.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test22.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test22.tig:7.11:Assigning wrong type to variable
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test23.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test23.tig:7.12:Assigning wrong type to variable
test/testcases/test23.tig:8.10:Assigning wrong type to variable
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test24.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test24.tig:5.2:subscript var on non-array type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test25.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test25.tig:5.2:field var on non-record type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test26.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test26.tig:3.3:Not int type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test27.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test28.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test28.tig:7.21:named type does not match expression
test/testcases/test28.tig:9.2:Variable does not exist!
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test29.tig correctly 1`] = `
"
val it = true : bool
intintarrtype2test/testcases/test29.tig:7.21:named type does not match expression
test/testcases/test29.tig:9.2:Variable does not exist!
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test30.tig correctly 1`] = `
"
val it = true : bool
int namesymbol:abtest/testcases/test30.tig:7.17:array type mismatch
test/testcases/test30.tig:7.14:named type does not match expression
test/testcases/test30.tig:9.3:Variable does not exist!
test/testcases/test30.tig:9.3:subscript var on non-array type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test31.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test31.tig:3.12:named type does not match expression
test/testcases/test31.tig:5.2:Variable does not exist!
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test32.tig correctly 1`] = `
"
val it = true : bool
intarraytytest/testcases/test32.tig:6.11:Initializing array with wrong type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test33.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test33.tig:3.10:Type undefined
test/testcases/test33.tig:3.10:Assigning record to not record type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test34.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test35.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test36.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test37.tig correctly 1`] = `
"
val it = true : bool
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test38.tig correctly 1`] = `
"
val it = true : bool
 namesymbol:int namesymbol:stringval it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test39.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test40.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test41.tig correctly 1`] = `
"
val it = true : bool
 namesymbol:int namesymbol:stringval it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test42.tig correctly 1`] = `
"
val it = true : bool
intrectype1stringarrtype1arrtype2arrtype3arrtype1test/testcases/test42.tig:27.15:Assigning wrong type to variable
test/testcases/test42.tig:28.15:Assigning wrong type to variable
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test43.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test43.tig:6.4:Not int type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test44.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test44.tig:5.16:named type does not match expression
test/testcases/test44.tig:9.2:Variable does not exist!
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test45.tig correctly 1`] = `
"
val it = true : bool
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test46.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test46.tig:4.16:named type does not match expression
test/testcases/test46.tig:6.2:Variable does not exist!
test/testcases/test46.tig:6.4:Not int type
test/testcases/test46.tig:6.4:Not int type
test/testcases/test46.tig:7.2:Variable does not exist!
test/testcases/test46.tig:7.4:Not int type
test/testcases/test46.tig:7.4:Not int type
val it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test47.tig correctly 1`] = `
"
val it = true : bool
 namesymbol:int namesymbol:stringval it = () : unit

"
`;

exports[`The textbook tests parses test/testcases/test48.tig correctly 1`] = `
"
val it = true : bool

uncaught exception Match [nonexhaustive match failure]
  raised at: src/typechecker/semant.sml:112.179
- 
"
`;

exports[`The textbook tests parses test/testcases/test49.tig correctly 1`] = `
"
val it = true : bool
test/testcases/test49.tig:5.18:syntax error: inserting  PLUS
test/testcases/test49.tig:5.10:Variable does not exist!
test/testcases/test49.tig:5.18:Not int type
test/testcases/test49.tig:5.18:Not int type
val it = () : unit

"
`;
